### PR TITLE
A grab bag of changes

### PR DESCRIPTION
**Pull request by randall**
Including:
 * Additional switches for setting TCP_NOTSENT_LOWAT and IP_TOS
 * Disable CORS headers by default (enable via --enable-cors)
 * Support for enabling h2c, HTTP/1.x and HTTP/3
 * Reduce some noise in logging of inconsequential errors
 * Upgrade to latest modules dependencies
 * Drop timing/response body from slurp handler (spec doesn't mention/need this)
 * Backwards incompatible: -config-port CLI switch renamed to -public-port

### CODE REVIEW

File: go/README.md [hunk 1]
```diff
@@ -14,22 +14,42 @@
...
   -key-file string
-    	key to use
+        key to use
   -listen-addr string
-    	address to bind to (default "localhost")
+        address to bind to (default "localhost")
   -public-name string
-    	host to generate config for
+        host to generate config for (same as -config-name if not specified)
+  -public-port int
```
Comment (hawkinsw) [inline go/README.md:43]: I am absolutely *not* worried about it, but I was just curious what made you decide to change?

File: go/announcer.go [hunk 2]
```diff
@@ -49,14 +50,15 @@ InterfacesLoop:
 		}
 	}
 	cfg := dnssd.Config{
-		Name:   "RPM Test Server",
+		Name:   fmt.Sprintf("RPM Test Server(%d)", port),
```
Comment (hawkinsw) [inline go/announcer.go:53]: Nice!

File: go/networkqualityd.go [hunk 3]
```diff
@@ -14,26 +15,42 @@ import (
...
+	"github.com/lucas-clemente/quic-go/http3"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
+	"golang.org/x/sys/unix"
 )
 
 var (
-	configPort = flag.Int("config-port", 4043, "The port to listen on for generating config responses")
-	publicPort = flag.Int("public-port", 4043, "The port to listen on for measurement accesses")
+	insecurePublicPort = flag.Int("insecure-public-port", 4080, "The port to listen on for measurement accesses")
```
Comment (hawkinsw) [inline go/networkqualityd.go]: Should we consider having different help text for this and `publicPort`?

File: go/networkqualityd.go [hunk 4]
```diff
@@ -14,26 +15,42 @@ import (
...
 	listenAddr = flag.String("listen-addr", "localhost", "address to bind to")
 
-	debug = flag.Bool("debug", false, "enable debug mode")
+	announce               = flag.Bool("announce", false, "announce this server using DNS-SD")
+	debug                  = flag.Bool("debug", false, "enable debug mode")
+	enableCORS             = flag.Bool("enable-cors", false, "enable CORS headers")
+	enableH2C              = flag.Bool("enable-h2c", false, "enable h2c (non-TLS http/2 prior knowledge) mode")
+	enableHTTP2            = flag.Bool("enable-http2", true, "enable HTTP/2")
+	enableHTTP3            = flag.Bool("enable-http3", false, "enable HTTP/3")
+	enableInsecureListener = flag.Bool("enable-insecure-listener", false, "enable insecure listener")
```
Comment (hawkinsw) [inline go/networkqualityd.go]: Should this be implied if the user gives `insecure-public-port`?

File: go/networkqualityd.go [hunk 5]
```diff
@@ -45,7 +62,7 @@ var (
 
 const (
 	smallContentLength int64 = 1
-	largeContentLength int64 = 4 * 1024 * 1024 * 1024
+	largeContentLength int64 = 32 * 1024 * 1024 * 1024
```
Comment (hawkinsw) [inline go/networkqualityd.go:61]: "This one goes to 11"

File: go/networkqualityd.go [hunk 6]
```diff
@@ -67,9 +84,29 @@ func setCors(h http.Header) {
 	h.Set("Access-Control-Allow-Headers", "*")
 }
 
+type handlers struct {
+	enableCORS bool
+}
+
+// BulkHandlers returns path, handler tuples with the provided prefix.
+func BulkHandlers(prefix string, enableCORS bool) map[string]http.HandlerFunc {
```
Comment (hawkinsw) [inline go/networkqualityd.go:90]: `bulkHandlers`?

File: go/networkqualityd.go [hunk 7]
```diff
@@ -116,44 +176,185 @@ func main() {
...
+					}
+					if setsockoptErr != nil {
+						return setsockoptErr
+					}
+				}
 
-	server := &http.Server{}
-	server.Addr = fmt.Sprintf("%s:%d", *listenAddr, *configPort)
-	server.Handler = mux
+				if tos > 0 {
```
Comment (hawkinsw) [inline go/networkqualityd.go:213]: Does it make sense to put this code in platform-specific file (the way that we have certain things in the client)?

File: go/networkqualityd.go [hunk 8]
```diff
@@ -116,44 +176,185 @@ func main() {
...
+				wg.Done()
+			} else {
+				if scheme == "https" {
+					if *enableHTTP3 {
+						log.Printf("Enabling H3 on %q", fmt.Sprintf("%s:%d", *listenAddr, port))
+						quicConf := &quic.Config{}
+						server := http3.Server{
+							Handler:    mux,
+							Addr:       fmt.Sprintf("%s:%d", *listenAddr, port),
+							QuicConfig: quicConf,
```
Comment (hawkinsw) [inline go/networkqualityd.go]: Could we put this straight inline?
```Go
QuicConfig: &quic.Config{},
```

File: go/networkqualityd.go [hunk 9]
```diff
@@ -275,39 +489,38 @@ func setNoPublicCache(h http.Header) {
...
+	}
+	if errors.Is(err, syscall.ECONNRESET) {
+		return true
 	}
 
-	js = append(js, '\n')
-	if _, err := w.Write(js); err != nil {
-		log.Printf("ERROR: Could not write response: %s", err)
+	switch err.Error() {
+	case "client disconnected":
```
Comment (hawkinsw) [inline go/networkqualityd.go]: Any way that we could find a constant variable for this? Seems like a "magic" value.

**Review by hawkinsw — APPROVED** [review]

File: go/announcer.go (hunk 2 above)
Comment (randall) [reply go/announcer.go:53]: Ultimately I'm not sure this is best here, especially when it comes to multiple listeners/ports/protocols. But I'll kick that can down the road.

File: go/README.md (hunk 1 above)
Comment (randall) [reply go/README.md:43]: which part? setting both `config-name` and `public-name` is a minor pain to think about when not running behind a proxy.

and the public-port was renamed to align with the addition of a second port (`insecure-public-port`)

File: go/networkqualityd.go (hunk 3 above)
Comment (randall) [reply go/networkqualityd.go]: Done.

File: go/networkqualityd.go (hunk 6 above)
Comment (randall) [reply go/networkqualityd.go:90]: This is for others adding into their own handlers (I'm working with groups who just want the data generation and not the rest of this).

The future here will be to split into package and command driver for the server.

File: go/networkqualityd.go (hunk 7 above)
Comment (randall) [reply go/networkqualityd.go:213]: Yeah, I'll make an another pass with that (for the big reorg). Will this stop you and your work?

File: go/networkqualityd.go (hunk 8 above)
Comment (randall) [reply go/networkqualityd.go]: Done.

File: go/networkqualityd.go (hunk 9 above)
Comment (randall) [reply go/networkqualityd.go]: Any suggestions? This is an unexpected error from the `http` package.

```
./net/http/h2_bundle.go:	http2errClientDisconnected = errors.New("client disconnected")
```

File: go/networkqualityd.go (hunk 4 above)
Comment (randall) [reply go/networkqualityd.go]: done.

File: go/networkqualityd.go (hunk 7 above)
Comment (randall) [reply go/networkqualityd.go:213]: actually, I split this up. To be refined at a later date (for now I have a !windows implementation and windows implementation.j)

**Pull request merged**